### ISSUE TITLE
Kasli specification

### ISSUE THREAD

**Issue by jbqubit** [m-labs/sinara#129]
@jordens is tracking design proposal/specification for Kasli on the wiki. Let's discuss PCB_kasli specification here. 

https://github.com/m-labs/sinara/wiki/Kasli

How Kasli might interface to peripherals is discussed in #99.

**Comment by jbqubit** [comment]
At teleconference today we discussed freezing Kasli electrical interface so that there is compatiablity with Metlino-VHDCI #99. Did I get this right? 
- Kasli uses +12V supply to Kasli
- 8 IDC connectors on Kasli
-  each IDC connector has 
    - 16 pins for general purpose IO (eg 8 LVDS)
    - 4 pins for general purpose IO or I2C or SPI
    - +12V supply, 100 mA max
    - +5V supply, 50 mA max
- Artix-7 FGPA 

@gkasprow Which student is working on Kasli? Please add to #19 and github.

**Comment by gkasprow** [comment]
I edited specification above.

**Comment by dtcallcock** [comment]
@jordens Can we have subheadings for each proposed mezzanine on the wiki?

Currently the wiki mentions:
- 16 TTL drivers
- 40 DAC channels
- CameraLink connector

But some other options were mentioned in #99 including SPI interface using LVDS over RJ45 and isolated TTLs.

I would also like to suggest a low cost synthesiser for driving non-critical AOMs/EOMs (ie, where you don't need deterministic phase control or sub-kHz resolution).  Would ideally also have pre-amp/switch/attenuators/filters so that only an external power amp is needed.  Could be based on the [ADF4351](http://www.analog.com/en/products/rf-microwave/pll-synth/fractional-n-plls/adf4351.html#product-overview) (an $8 synth-on-a-chip) like the [SynthNV](https://windfreaktech.com/product/rf-signal-generator-and-power-detector/) is.

**Comment by jordens** [comment]
@dtcallcock sure. Do you want to edit that yourself?

**Comment by dtcallcock** [comment]
@jordens I have had a go at reorganising the wiki page based on these discussion threads and the work done so far by @gkasprow and @marmeladapk.  You guys should fill in any blanks and correct where I've misunderstood what you're up to.  I'm also a little unsure when exactly we put things on the wiki.  Do we do it right away and then discuss in the issue or suggest in the issue and move to the wiki one we've reached some kind of consensus?

I also have a few questions:

- The /ARTIQ_ALTIUM/PCB_3U_BNC folder contains two .pdf output files, one a few weeks older than the other.  Is this a mistake?

- What voltages go over the IDC?  Some schematics have +5V and some have +3.3/+12V.  Also what are the current specs?

- What is the 5th RJ45 on 3U_RJ45 going to be used for?

- Can we use thread-lock barrel connectors on the boards?  These are much more robust in a lab setting and are backwards-compatible with non-locking plugs for those that don't care.  Example parts are [connector](http://www.digikey.com/products/en?keywords=sc1472-nd) and [plug](http://www.digikey.com/product-detail/en/switchcraft-inc/761K/SC1148-ND/241919).

- Should the barrel connector input on PCB_3U_VHDCI_Baseboard have some minimal EMI filtering?

**Comment by gkasprow** [comment]
I deleted wrong pdf file.
We use 12V and 3.3 for EEPROM readout. It is available even when the box is not powered.
The fifth R45 was meant as I2C link for slow devices, but I don't have strong opinion about it.
Yes, we can use such connectors. Just add an issue for Michal.
We can install CMC and EMI filter but such should be already in the power supply.

**Comment by dtcallcock** [comment]
> We can install CMC and EMI filter but such should be already in the power supply.

The power supply was originally required to be "just some random wall wart".  In my experience these are often horrifically noisy and have poor regulation.  Should we add circuitry to deal with that or just change the spec to "reasonably clean isolated +12V supply".

>The fifth R45 was meant as I2C link for slow devices, but I don't have strong opinion about it.

I'd probably rather leave this off due to I2Cs lack of robustness to bus capacitance changes.  SPI will be more better for off-board devices.  If remote I2C devices are really desired though, we could utilise bus buffers like the [PCA9600](http://www.nxp.com/products/interfaces/ic-bus-portfolio/ic-voltage-level-translators/dual-bidirectional-bus-buffer:PCA9600).

There is also a high likelihood of accidentally plugging in one of the adjacent LVDS cables or vice-versa, though this could be avoided by using a different connector. 

> I deleted wrong pdf file.

The .pdf there now is the one that still has the old 5V power rail.

**Comment by gkasprow** [comment]
Indeed, there is wrong schematic version still with 5V bus.

**Comment by gkasprow** [comment]
fixed

**Comment by jordens** [comment]
@dtcallcock looks good. Thanks! I am all on board with the additions/changes and have added a few details.

* I am okay with saying "reasonably clean 12V". But IME there are plenty of examples where standard wall wart supplies have been without problems. E.g. in the USRP. I think we can aim to replicate that.
* Supply-wise I don't care. Grep should probably choose what's best/easiest.
* The thread lock barrel connectors (or similar Bajonett style things) are a good idea.

**Comment by hartytp** [comment]
Just to check: the current plan is that each Kasili communicates with/receives time from the root Metlino via a fibre link (one fibre per Kasili), and not via a backplane, right?

**Comment by jordens** [comment]
Yes. The VCO/DDS synthesizer extensions would receive their reference clock individually.

Another issue is experiments where the set of Kasli extensions is sufficient. In those experiments Metlino (plus Rack, clock RTM) seems pointless and too powerful/big and requiring a separate box/piece of hardware just to be able to talk to Kasli is troublesome.

**Comment by hartytp** [comment]
> Yes

@jordens Good, that seems like the correct solution to me. IIRC, the Metlino currently has 3 SFPs, which is extensible to 7 by adding an FMC-SFP board. However, given the number of different functions currently planned for Kasili, I can imagine using these all up pretty quickly. What's the plan for extending this further?

> The VCO/DDS synthesizer extensions would receive their reference clock individually.

Sure, if that's useful, although IMO it's a bit overkill: IIRC, the recovered clock should actually be pretty good noise-wise and, I assume that one be better off use Sayma for any ultra-low noise purposes. 

> Another issue is experiments where the set of Kasli extensions is sufficient. In those experiments Metlino (plus Rack, clock RTM) seems pointless and too powerful/big and requiring a separate box/piece of hardware just to be able to talk to Kasli is troublesome.

What's your plan for this use case?

**Comment by dtcallcock** [comment]
> >The VCO/DDS synthesizer extensions would receive their reference clock individually.

> Sure, if that's useful, although IMO it's a bit overkill: IIRC, the recovered clock should actually be pretty good noise-wise and, I assume that one be better off use Sayma for any ultra-low noise purposes.

I can see how you might want this as an option for a DDS.  For the synth though, can the Kasli FPGA just provide the clock over one of the LVDS pairs?  I'm not entirely sure what the Artix clock output capabilities are though.

**Comment by jordens** [comment]
* You get a bunch of SFPs with each Sayma. You can also hook up another (slave) Metlino to the (master) Metlino. And we can daisy chain Kaslis ad infinitum. All those require DRTIO switch support and consensus about one preferred option that would be implemented.
* Distributing the clock from Kasli to the extension boards is tricky. There is little space on the frontpanel, doing it inside makes assembly even harder than it is now (now you already have to disassemble the crate when you change/add/remove extensions), and there is little space on the PCB for SMA connectors.
* I would be afraid to run a reference clock over uncontrolled impedance, unshielded, IDC connectors, even for a simple Synth. We would also need a clock fan-out as the Artix outputs are very noisy. I am fine with putting an XO on the extension boards as a low complexity alternative to a reference clock.

**Comment by hartytp** [comment]
> You get a bunch of SFPs with each Sayma. You can also hook up another (slave) Metlino to the (master) Metlino. And we can daisy chain Kaslis ad infinitum. All those require DRTIO switch support and consensus about one preferred option that would be implemented.

Good, that works for me. Is the plan for each Kasli to have 2xSFP to allow daisy chaining then?

**Comment by jordens** [comment]
Right now I don't see a different use for them.

**Comment by hartytp** [comment]
> I would be afraid to run a reference clock over uncontrolled impedance, unshielded, IDC connectors, even for a simple Synth. 

Depends on what the reference clock frequency is. For, say, a 100MHz reference clock frequency, routing board-board via pin header will be fine.

IIRC, the AD9912 has a pretty decent clock multiplier (PLL w/ integrated VCO), so there's no need to supply a GHz clock directly, unless one wants to push the noise performance (in which case, the Sayma might be a better choice anyway). Otherwise, one can always use something cheap like an AD9576 to do the same job.

IMO, it'd be great if these Kasli extension boards could be used without needing an external GHz frequency source...

**Comment by hartytp** [comment]
@jordens My previous post was ambiguous, I meant "Is the plan for each Kasli to have 2xsSFP (which would allow them to be daisy chained)?" Answer: "yes", so I'm happy with this plan.

**Comment by hartytp** [comment]
> We would also need a clock fan-out as the Artix outputs are very noisy. 

Won't there be a spare output from the Si5324 we can use to supply the reference clock?

**Comment by jordens** [comment]
Yes, there are currently two SFP cages on Kasli.
Those impedance discontinuities will give us reflections which can quickly lead to double clocking leading to the need for filters etc. I would not want to go through that again.
Yes. The SI5324 has one other output. But there are eight extension cards.

**Comment by hartytp** [comment]
> Those impedance discontinuities will give us reflections which can quickly lead to double clocking leading to the need for filters etc. I would not want to go through that again.

IME, this isn't too much of an issue for clocks in the 50-100MHz range. Besides which, aren't you planning to run an SPI bus at something like 50MHz over this pin header?

IIRC, the AD9912 has a max PLL divider ratio of 66, so the reference clock can be as low as 15MHz while still running at max output frequency. I really don't see any issues with reflections at that kind of clock speed.

**Comment by hartytp** [comment]
> Yes. The SI5324 has one other output. But there are eight extension cards.

Sorry, I'd misread a previous post, and was talking cross-purposes here. Of course you need a fanout buffer for this...

**Comment by jordens** [comment]
SPI/I2C is typically de-glitched AFAIK (at least our SPI slave is de-glitching). And with SPI you can always go slower if it doesn't work.
And for a square clock it's not the frequency that determines the reflections. It's the edge speed and the bandwidth of the input buffer at the other end. A 15 MHz clock with the same fast edges as a 1 GHz clock from a good digital fanout will give you the same distorted edges and potential double-clocking at the (fast) input of the AD9912.
We've had this at 50 MHz over much shorter connections and better-than-IDC connectors.
I'd prefer to be able to choose between standard on-board XO and good external input.

**Comment by hartytp** [comment]
> And for a square clock it's not the frequency that determines the reflections. It's the edge speed and the bandwidth of the input buffer at the other end. A 15 MHz clock with the same fast edges as a 1 GHz clock from a good digital fanout will give you the same distorted edges and potential double-clocking at the (fast) input of the AD9912.

Sure, but if your fundamental is low, you can use a slower fanout buffer (w/ hysteresis if necessary)/lower drive current or add some RC to reduce the edge time without causing problems. 

> I'd prefer to be able to choose between standard on-board XO and good external input.

IMO, it's nice to have absolute accuracy from a PLL without having to supply an external (SMA) reference. But, I don't feel too strongly about this.

**Comment by dtcallcock** [comment]
> I am fine with putting an XO on the extension boards as a low complexity alternative to a reference clock.

A decent XO will get us <1ppm which will be fine for a >1kHz resolution synth.

> The SI5324 has one other output. But there are eight extension cards.
>Distributing the clock from Kasli to the extension boards is tricky. There is little space on the frontpanel, doing it inside makes assembly even harder than it is now (now you already have to disassemble the crate when you change/add/remove extensions), and there is little space on the PCB for SMA connectors.

Could we run that spare SI5324 output to a single SMA?  Then people can use it as their 'external reference' if they so choose.  The distribution problem (for >1 extension card) still has to be solved whether they use that or another source they provide.

**Comment by dtcallcock** [comment]
Another option for clock distribution is to use the backplane.  Cards could revert to using the IDC (with no clock) if no backplane is present.  

Is there currently a plan for the backplane?  The Kasli and VHDCI carriers are currently planned to have a DIN 41612 C, 96 position connector.  This would presumably make them mechanically compatible with VME J1 backplanes.  Is the idea to somehow use those or just design a custom one?  With the 96-position connector it would be trivial to wire the equivalent payload of the IDC headers (8xLVDS pairs, I2C, power) plus a clock to 4 slots in a star topology.

**Comment by hartytp** [comment]
> A decent XO will get us <1ppm which will be fine for a >1kHz resolution synth.

<1ppm is pretty good for an XO -- most of the ones I've seen are more like 10ppm initial accuracy and then ~1ppm/year ageing. But, I take your point, this won't be an issue.

Another potential advantage of using an XO, rather than distributing the clock via ribbon cable, is that it makes it easier to run these extension boards off a standard microprocessor board like an Arduino.

**Comment by gkasprow** [comment]
There is no standard backplane on the market, so one would have to design it.
In case of VHDCI, we would need dedicated adapter from VHDCI to the backplane.
For the moment we can just add SMA or MMCX clock output for extension boards that need it.
DO you really need to connect 8 extension  boards with clock input?
Of course we can put clock splitter at the SIlabs output to generate 8 copies of the clock.

**Comment by hartytp** [comment]
@jordens When Kasli is mounted in [an enclosure](http://schroff.pentair.com/en/schroff/plug-in-units---frame-type---kits/3074457345616861429) how do you plan to bring out the ribbon cables for the daughter boards? Front panel connectors?

**Comment by jordens** [comment]
@hartytp they run inside the crate between the cards. See the screenshots on the wiki page.

I am ok with routing the other SI5324 output to a front panel SMA.

I would enable the use of the backplane in the same way as the ribbon cables. I.e. have four extension connectors routed to IDC/ribbon *and* a pin group on the backplane/DIN. The backplane would then do a star topology to up to four the extension card. But I would not require the use of the backplane as it becomes tedious to design new backplanes for small (single extension) enclosures.

**Comment by hartytp** [comment]
@jordens The Wiki mentions using Kasli in standalone mode in a frame enclosure, so no rack in this case. Or, is the plan that Kasli never gets used outside (at least a small) small rack?

I assume that when Kasli is in a rack, it just gets a front panel, rather than a full frame enclosure (otherwise, you need a hole in the enclosure to get ribbon cable out, which is inconvenient).

**Comment by jordens** [comment]
But the enclosures are "similar" to racks in that you can slide in the boards the same way and screw the frontpanels into the front of the enclosure. You would slide Kasli and one (or more) extensions into a small enclosure after having connected them with ribbon cables. The only real difference is the size and the modularity of the enclosure vs the rack.

**Comment by hartytp** [comment]
@jordens Okay, so is the plan is to design enclosures (or, rather, front panels for enclosures) that can fit Kasli + some (1? 4?) extension boards?

**Comment by jordens** [comment]
Yes. Frontpanels will be made. They are per extension/Kasli.

**Comment by dtcallcock** [comment]
>I would enable the use of the backplane in the same way as the ribbon cables. I.e. have four extension connectors routed to IDC/ribbon and a pin group on the backplane/DIN. The backplane would then do a star topology to up to four the extension card. 

Sounds good.  For the VHDCI carrier the 4 IDC/ribbon connectors connected to the backplane should all come from the same VHDCI connector (so Sayma FMC-VHDCI isn't limited to using two backplane slots).   For the Kasli carrier, is there enough FPGA IO that the backplane could have its own lines (thus in principle allowing for 12 extensions)?

> But I would not require the use of the backplane as it becomes tedious to design new backplanes for small (single extension) enclosures.

Yes, the IDC header should be provided on all extensions too.

How do we feel about adding a clock pair to each backplane pin group as well?  On the VHDCI carrier these would be driven by a front panel SMA input.  The Kasli would drive them with its recovered clock.  Extensions needing a clock would then have a switch to choose between this one and one provided on a (front panel?) coax connector.

**Comment by hartytp** [comment]
If we do put a CLK_IN SMA on the VHDCI carrier front panel, used to drive the back plane, could we also add a buffered CLK_OUT SMA to allow daisy chaining? If this isn't easy (e.g. due to space considerations) then don't bother.

**Comment by gkasprow** [comment]
we already have on he front panel:
2xSFP
micro USB for UART and JTAG
SATA connector
we can still fit 2 SMA connectors
look here

![obraz](https://cloud.githubusercontent.com/assets/4325054/22487633/6b4e4aae-e80e-11e6-815e-81caf5c6ff8c.png)

**Comment by jordens** [comment]
For a few projects we want to run Kasli stand-alone. I.e. Ethernet over SFP, coredevice and DRTIO master on Kasli. That means there won't be a recovered clock but one that comes from the frontpanel.

If we can have one SMA on the front panel that's great. That would be either an output for the recovered clock (in DRTIO satellite mode) or an input for the RTIO reference clock (in Kasli stand-alone mode). I am ok if that is selected by capacitor placement.

If we really consider it a high value goal to distribute clock (over the backplane), and if we can limit the complexity to a single simple dumb clock fan out that does not need configuration, and if there is space on the panel, I am ok with that second SMA being used as an input to an independent clock distribution from the Kasli board to the backplane. It is getting a bit tight with the pins on the backplane connector (`96 [available] = 4 [extensions]*(2*8 [lvds IO] + 2 [12V] + 1 [3.3V] + 2 [I2C] + 2 [lvds clock] + 1 [GND])` is cutting it way too close with GND. We should probably limit the backplane to three extensions.

**Comment by gkasprow** [comment]
We can install SIlabs chip with fanout IC.
Silabs has 2 inputs so one can be routed to the front panel
One output goes to the FPGA, another to the fanout IC.
Fanout IC  then to second SMA on a front panel, to the backplane and to FPGA
Second input of the fanout could be connected to input SMA as well.
In this way we could use it in daisychain configuration as well, without SIlabs in the loop.

**Comment by hartytp** [comment]
@jordens To confirm what it says on the Wiki atm: neither of Kasli DDS or VCO will have analogue inputs, so no servos will be possible without additional hardware, right?

**Comment by jordens** [comment]
@gkasprow That sounds good.
@hartytp Yes. But they have means to adjust power. We plan to implement (slow) servos with additional hardware (digital distributed photodetectors).

**Comment by hartytp** [comment]
> We plan to implement (slow) servos with additional hardware (digital distributed photodetectors).

Nice. In a sufficiently compact form-factor, that could be a really great thing to have around the lab. 

Would that connect to Kasli using I2C or something like that? Do you have a draft proposal/writeup/specs or an estimated time-line for this?

**Comment by jbqubit** [comment]
+1 to limited clock distribution over backplane

Seven days ago @jordens said

>You get a bunch of SFPs with each Sayma. You can also hook up another (slave) Metlino to the (master) Metlino. And we can daisy chain Kaslis ad infinitum. All those require DRTIO switch support and consensus about one preferred option that would be implemented.

Some questions about daisy chaining. Example daisy-chain: ```Metlino-Kasli-Kasli-Kasli```
- Current plan is for 2 SFP per Kasli. That leaves no free SFP for fiber-based connection to a remote peripheral or to use the transceiver photo-detector for photon detection. Consider removing the front-panel SATA and replace with more SFP(s). Support of  SATA, SFP, ribbon and backplane seems an unnecessary proliferation of interconnect options. 
- Google says there are commercially available SFP to SATA. 
    - https://shop.trenz-electronic.de/en/TE0424-01-SFP-2-SATA-Adapter
- What do you anticipate for latency per Kasli in my example? Will it be deterministic? 
- Is there a proposed [DRTIO switch](https://github.com/m-labs/artiq/issues/619) implementation? 
- Please restate question for which you'd like consensus.

**Comment by jordens** [comment]
@jbqubit Looking at Sayma AMC again, there are no transcievers connected to the Fayma AMC FMC. There are not enough to do everything. The connections available for designation as downstream DRTIO could be the two SFP, the two SATA and the two FAT_PIPE links.

Using the transcievers for photon detection is tricky (very) because you have a (strong) DC bias and lots of low frequency energy. I.e. they can't be AC coupled, which is what SFP modules and adapters tend to do internally. Apart from that I would not like to squeeze that functionality into the use cases Kasli is targeting. Use the SATA port or the second SFP port, or a port on Sayma (which also has faster transcievers) for that. The SATA connector is convenient and adequate for daisy chaining within the same rack. This is completely orthogonal to the IDC ribbon and backplane connections. I don't understand why you bring them up here.

Latency will be deterministic. Kasli is a DRTIO satellite.

There is a DRTIO switch implementation path that we have discussed, also on IRC and indirectly in that thread. Yes.

When we write the gateware for Sayma and Metlino, we need to decide what's in the FMC slot(s), what that FMC is used for, and what the SATA/SFP ports are used for. TTLs? Transcievers? Photon tagging? DRTIO downlinks? DRTIO uplinks? ADC FMC? DAC FMC? CameraLink?

@hartytp The photodetectors would likely be developed by PTB. I'll try to convince them to go open hardware with them. I2C most likely, otherwise "typical" Ion trap photodetectors, nothing more specific yet. Timeline goal is to be coincident with the rest of Kasli, maybe a bit later, maybe a bit earlier.

**Comment by dtcallcock** [comment]
> It is getting a bit tight with the pins on the backplane connector (96 [available] = 4 [extensions]*(2*8 [lvds IO] + 2 [12V] + 1 [3.3V] + 2 [I2C] + 2 [lvds clock] + 1 [GND]) is cutting it way too close with GND. We should probably limit the backplane to three extensions.

There are opportunities for more grounds:
- 3.3V is low current (20mA per extension) so we can use a single pin, giving 3 more grounds.
- With a backplane, the 12V will come directly to the backplane and not via the barrel connector on the carrier, right?  If so then 2*12V for Kasli carrier power should be sufficient, giving 6 more grounds.

At this point we'd have 13 grounds.  There is also the less desirable option of:
- Clock fanout and/or I2C switch on the backplane.  Up to 12 more grounds this way.

**Comment by hartytp** [comment]
> @hartytp The photodetectors would likely be developed by PTB. I'll try to convince them to go open hardware with them. I2C most likely, otherwise "typical" Ion trap photodetectors, nothing more specific yet. Timeline goal is to be coincident with the rest of Kasli, maybe a bit later, maybe a bit earlier.

Thanks for the update. Good luck! IMO, this is quite important, since the DDS boards are significantly less useful to us without some kind of servo...

**Comment by dtcallcock** [comment]
> We plan to implement (slow) servos with additional hardware (digital distributed photodetectors).

I was just thinking of starting a discussion of this.

> I2C most likely

One of the problems of these 'slow' servos is that they are often feeding back on few-us pulses.  Thus if a relatively slow bus like I2C is being used, a fast acquisition trigger or sample and hold should also be implemented.

> otherwise "typical" Ion trap photodetectors

I quite like the way Moglabs have implemented this.  They use a firewire cable that contains +/-12V and a shielded twisted pair for the differential photodiode signal.  See p84 of [this manual](http://www.moglabs.com/uploads/2/4/2/1/24212474/moglabs_dlc_manual_rev916.pdf).

**Comment by dtcallcock** [comment]
> When we write the gateware for Sayma and Metlino, we need to decide what's in the FMC slot(s), what that FMC is used for, and what the SATA/SFP ports are used for. TTLs? Transcievers? Photon tagging? DRTIO downlinks? DRTIO uplinks? ADC FMC? DAC FMC? CameraLink?

Using the FMC-VHDCI card and VHDCI carrier would presumably allow reuse of Kasli extensions and gateware for some of the above (TTL/ADC/DAC/CameraLink etc).

**Comment by jbqubit** [comment]
>The SATA connector is convenient and adequate for daisy chaining within the same rack.

How does this enable: ```Metlino-Kasli-Kasli-Kasli-...``` Two identical ports are required for daisy-chaining. 

What's wrong with adding more SFP on the front panel and using a SFP-SATA adapter as required? 

In what circumstances do you see SATA as preferred to SFP? SFP transceivers are $20/pair. 

>When we write the gateware for Sayma and Metlino, we need to decide what's in the FMC slot(s), what that FMC is used for, and what the SATA/SFP ports are used for. TTLs? Transcievers? Photon tagging? DRTIO downlinks? DRTIO uplinks? ADC FMC? DAC FMC? CameraLink?

This broad topic is discussed in https://github.com/m-labs/sinara/issues/139. I created a new Issue for FMC #148. Let's not discuss this here.

**Comment by jbqubit** [comment]
@dtcallcock comment moved to #148.

**Comment by jordens** [comment]
@dtcallcock 
Firewire is a pretty atrocious connector. There seem to be systematically bad experiences with stability and reliability e.g. in the SRS shutters. But anyway: Yeah, similar it would be. Modulo the question whether one digitizes at the frontend or later.

Since this is DRTIO, if you can trigger over the bus (I2C/SPI) you can time the sampling so that it meets your window without any additional signals.

@jbqubit 
Kasli has two SFP, you can therefore do `Metlino--Kasli--Kasli--Kasli--` as much as you want.
SATA is preferred over SFP for short distances because it is less complicated, more robust, easier to get the right components, and cheaper.

**Comment by hartytp** [comment]
> I was just thinking of starting a discussion of this.

Good idea. Probably worth its own issue though?

**Comment by gkasprow** [comment]
Guys, we can fit 4 SFPs on Kasli + USB and 2 SMA
And for SFPs there are copper cables available and several extensions like 10/100/1000T Ethernet, fibre, SDI, HDMI (cameras). No reason to go for SATA
![obraz](https://cloud.githubusercontent.com/assets/4325054/22524605/36e9e6e0-e8c4-11e6-8ab1-90dbc509a554.png)

**Comment by gkasprow** [comment]
We can use numerous existing FMC extensions in Sayma directly.

Fast ADCs, DACs (up to 16 channels)

Fast IO (32x TTL)

TDC, DTC, programmable delay

LVDS IO 

Camera link

In the future we can build  deterministic computing module with floating point support

**Comment by gkasprow** [comment]
For daisy chaining we can use copper sfp patchcords

In addition, SFP has module detection mechanisms.

You can even hook up camera link to SFP using adapter.

**Comment by jordens** [comment]
@gkasprow Pretty sure that the SMA cables interfere with the SFP modules and the USB cable. That's much too close. Could we please do at most 3 SFP, 1 USB, 2 SMA?
I am also worried about PCB real estate for all the IDC connectors, FPGA power supplies, SDRAM, flash, SFP cages, clock fanout. I'd like to focus on what's important and needed now and put all those things that could be possible in addition a bit lower on the topic list.

**Comment by gkasprow** [comment]
Don't worry about real estate - there is enough room
we can try with micro USB
![obraz](https://cloud.githubusercontent.com/assets/4325054/22528659/8ec818d6-e8d4-11e6-83d6-ace13a5ec616.png)

or as you wrote, stay with 3 SFPs and route one to onboard SATA.

**Comment by hartytp** [comment]
That particular combination of SMA USB SMA looks somewhat alarmed and panicked....

other than the visuals though, looks good to me ;)

**Comment by gkasprow** [comment]
The only reason is to separate 2 SMAs :)

>

**Comment by sbourdeauducq** [comment]
We can also remove one SFP to make some space and potentially ease PCB routing a bit.

**Comment by klickverbot** [comment]
To entertain the quite probably rather premature front panel discussion for a little longer: Just by eye, the SMAs look rather close to the micro USB jack – are you sure two standard plugs would fit at the same time? I imagine one might want to use both an external clock and UART/JTAG for debugging.

**Comment by jordens** [comment]
@jbqubit In the current state of things we will be funding Kasli and while we acknowledge and appreciate your recommendation on this topic, we will go a different route.
We'd like easy and pain-free access to all front panel element for regular-sized (i.e. fat) fingers. That means reducing the density.

**Comment by hartytp** [comment]
@jordens To be explicit then, what front panel connections do you plan to include? (2xSMA, 1xUSB, 3xSPF?)

Presumably, at some point in the future we will make a cheap SFP hub board to go with Kasli. IMO, so long as this is planned, there's no need to over-stuff Kasli's front panel with SPFs.

**Comment by jordens** [comment]
@hartytp yes. 3 SFP, 1 USB, 2 SMA. 3 SFP already gives us more than sufficient scalability with only logarithmic DRTIO tree depth and plenty of unused transcievers at the leaves.

A "SFP hub" would need to be a full DRTIO switch.

I am confused by the apparent worry of not being able to scale due to lack of SFP ports. To me this seems unfounded and I value a good and usable mechanical design very highly. What is the biggest ARTIQ hardware tree that you can reasonably think of and why would one need more that the currently planned number of DRTIO downstream ports in Metlino/Sayma/Kasli?
Somebody insisted that Metlino/Sayma would be designed to fill an entire 12-slot crate and the transcievers were routed to the backplane accordingly. That gives you for a single Metlino: 12 Saymas, 27 DRTIO downstream ports from the µTCA crate, enough ports to drive 7 full VHDCI breakout crates, and up to 33 Kaslis with their crates, everything at a maximum DRTIO tree depth of 2.

**Comment by gkasprow** [comment]
There is an option to use 6HP panel (instead of 4HP) and install stacked SFPs so we would gain plenty of space.

**Comment by hartytp** [comment]
> @hartytp yes. 3 SFP, 1 USB, 2 SMA. 3 SFP already gives us more than sufficient scalability with only logarithmic DRTIO tree depth and plenty of unused transcievers at the leaves.

Works for me.

**Comment by hartytp** [comment]
> There is an option to use 6HP panel (instead of 4HP) and install stacked SFPs so we would gain plenty of space.

IMO, there's no need to do this. 3xSPF should be plenty, and it's nicer to keep Kasli compact and simple.

**Comment by jordens** [comment]
3 SFP is plenty IMHO, I'd very much appreciate an internal SATA connector (standard, i.e. "computer" pinout, not swapped "disc drive") for the 4th transciever.

**Comment by gkasprow** [comment]
Add this requirement to spec on wiki.

**Comment by jbqubit** [comment]
3 SFP, 1 internal SATA, 1 USB, 2 SMA

Sounds good to me. @jordens Please edit wiki.

**Comment by jbqubit** [comment]
Will DMA streaming (from DRAM on Kasli) be available to Kasli peripherals? eg Zotino

**Comment by sbourdeauducq** [comment]
Yes, DMA is hardware-independent.

**Comment by jbqubit** [comment]
@jordens Please update Kasli specification on wiki to reflect details fleshed out in this Issue.

**Comment by hartytp** [comment]
@jbqubit @gkasprow Did you guys decide how you'd like the Kasli backplane to be wired up? I'm not entirely clear what topology was suggested (agreed on?). If anyone has thoughts, could they put them on the Wiki, please?

**Comment by gkasprow** [comment]
We specified with @marmeladapk that remaining LVDS lines from FPGA on Kasli will be routed to 96pin connector. The signals will follow signalling available in IDC connectors. We should be able to wire 3 or 4 x8 LVDS pairs + I2C buses.
Kasli will also distribute copies of the Si5325 clock in point to point configuration.
It may be used for two kinds of applications:
- requiring high speed signalling where IDC ribbon cannot do the job.
- some not yet defined applications that would benefit from B-LVDS  or MLVDS signalling where LVDS pair runs across the backplane and several boards can talk to it. One can achieve 100Mb/s operation.
In this way we can distribute i.e. high speed multidrop SPI over LVDS and with one Kasli connect several devices to it. We can also use the backplane to synchronize the modules.
To answer your question - no, the backplane is not yet defined. We simply lack use cases.

**Comment by dtcallcock** [comment]
I added my suggested  backplane connector allocation from earlier to the wiki.  

There is also the issue of wiring the 96pin connector on the VHDCI carrier.  My preference would be to route all four IDCs from one VHDCI connector to it (rather than for example two IDCs from each VHDCI).  Then you have to option of running two VHDCI carriers and two backplanes off a single Metlino.

**Comment by dtcallcock** [comment]
Also, should we add an SMA front panel input and clock fanout chip to VHDCI carrier so it can drive the backplane clocks?

**Comment by gkasprow** [comment]
We can place IDC connector routed to 96pin connector that follows Kasli connectivity. In this way we could either use 8 IDCs or route unused ones to 95pin connector (DIN 41612) using short IDC jumper.
To make both boards fully compatible, we would need to addd clock fanout as well.
We can implement it in next revision of VHDCI board.

**Comment by hartytp** [comment]
Thanks for that guys.

Maybe I'm being a little slow here, but I'm still not entirely clear about what exactly we're trying to achieve with this backplane. I can think of three potential reasons to have a backplane:

a. Reducing wiring complexity (no ribbon cables to modules, no external SMAs needed to distribute clocks)?
b. For high-speed digital communication that can't be sent over the ribbon cable
c. Some kind of multi-point bus link to allow communication between extension modules without going through the carrier

(a) Is potentially quite nice, but IMO probably not worth the effort of designing and maintaining a backplane. Also, as previously discussed, if we use 96 pin connectors then a carrier can only drive 4 extension modules, which isn't convenient: each carrier has 8 expansion headers, so the BP wastes some connectivity; and standard racks will hold far more extension modules than this so we end up with empty slots or multiple backplanes per rack or something. 

(b) and (c) are what @gkasprow suggests. But, AFAICT, no one has suggested a use case for them yet. We can't (shouldn't) specify a backplane for undefined purposes. So, if this is our motivation, maybe we should just leave the 96-pin connectors NC until we have a more concrete use case to design the BP around.

Whatever the answer, it would be good to agree on our motivation/goals for this, and to put that on the Wiki...

**Comment by gkasprow** [comment]
Actually, it's hard to fit more than  32 LVDS pairs (equivalent of 4 IDC connectors). They already would occupy 64pins.
Add 6 or 8 clock outputs and we have 80 pins. Add 4* I2C and we have 10 pins left for grounds and supplies which is absolute minimum.

**Comment by dtcallcock** [comment]
If a) was the priority but only worth doing if we could drive all 8 extensions then we could switch to a 160 pin connector on the Kasli/VHDCI carrier (would also need the I2C switch and/or clock fanout moving to backplane).  I think this only makes sense if the BNC/SMA/RJ45 extensions were also made full length to connect to the backplane.  As it is those can only be connected via IDC so in most cases I would anticipate a mix of up to four full length cards (DDS/Synth/ADC/DAC etc) connected via the backplane and the rest of the connectivity taken up by the simpler IDC-only extensions.  

On the other hand, if we thought a) wasn't worth the trouble but b) and c) were the real worry, we could drop the backplane and stick a bunch of SATA connectors (or similar) on the back of Kasli instead.  The hypothetical extension requiring high speed connections would still use the IDC for power/I2C/slowstuff but could also be connected to SATAs as needed for high speed links.  SATA would also be used for c).

We could also just wait until we're got our hands on the prototypes before deciding what to do here.  I think some hands-on use will inform how much effort we are willing to put in to achieve a).  For reference in moving from the previous the current generation of NIST hardware we _did_ consider it worth the effort to move from IDCs to a backplane.  Also use cases for b) and c) might arise in the meantime.

**Comment by hartytp** [comment]
**Option a**

> @dtcallcock: I think some hands-on use will inform how much effort we are willing to put in to achieve a). For reference, in moving from the previous to current generation of NIST hardware we did consider it worth the effort to move from IDCs to a backplane.

ack

> I think this only makes sense if the BNC/SMA/RJ45 extensions were also made full length to connect to the backplane. 

Agreed: IMO, if we're using a backplane to cut down on wiring then all boards should connect directly to it without ribbon cable. This will require a minor revision to some boards.

> @gkasprow: Actually, it's hard to fit more than 32 LVDS pairs (equivalent of 4 IDC connectors).

> @dtcallcock: If a) was the priority but only worth doing if we could drive all 8 extensions then we could switch to a 160 pin connector on the Kasli/VHDCI carrier

Agreed: if we're planning to use a backplane for this then we will probably want to switch to a higher density connector.

There's also the question of how many extension modules we can fit in the rack. A standard rack is something like 84HP wide. I assume that we'll have "wide" extensions like the BNC breakout, DACs, etc which will be 12HP, and "narrow" extensions like the RJ45 breakout, at 4HP? So, do we want a single carrier per rack (e.g. either VHDCI _or_ Kasli in a rack), and then up to 8 extension cards? A rack can then fit up to 6 wide extensions + carrier. All 8 BP slots can be used if narrow extensions are used as well. Obviously, this will leave a lot of unusable space if one populates the rack with only narrow extensions like the RJ45 breakout, but IMO we can live with that.

----

> @dtcallcock: On the other hand, if we thought a) wasn't worth the trouble but b) and c) were the real worry, we could drop the backplane and stick a bunch of SATA connectors (or similar) on the back of Kasli instead. The hypothetical extension requiring high speed connections would still use the IDC for power/I2C/slowstuff but could also be connected to SATAs as needed for high speed links. SATA would also be used for c).

Exactly, I think this was @jordens motivation for putting some SATA connectors on Kasli (inside the rack, rather than on the front-panel).

---

> @dtcallcock: We could also just wait until we're got our hands on the prototypes before deciding what to do here. I think some hands-on use will inform how much effort we are willing to put in to achieve

Agreed, it's probably better to wait until we've had a play with the hardware than to rush into specifying a backplane interface without due thought.

However, obviously the longer we put this decision off, the more hardware we're likely to design in the mean time (ADCs DACs SPI I2C etc). That means that we'll have to alter more boards to fit the backplane when a standard is agreed. IMO, so long as we can agree on what we're trying to achieve now then we should be able to specify it now.

It sounds to me like:

* Out chief motivation for a backplane is (a): cutting down on wiring clutter
* Any BP that can works for (a) will also work for (b) and (c), so we get those for free, even though they're not our main aim with the BP.
* High-speed interfaces (b) will additionally be supported by SATA on Kasli (not VHDCI)

Thoughts?

**Comment by gkasprow** [comment]
We decided to go for IDCs to enable out of rack operation of Kasli extensions.
If we switch to backplane connectivity we would have to develop dedicated custom backplane for every combination of extension modules.
Keep in mind that these modules may have different size so there will always be wasted some slots on the backplane.
IDCs can be low or high performance. You can easily run 2.5Gbit (PCIe) data over such cable so there should be no issues with DDS modules.

**Comment by sbourdeauducq** [comment]
Also Kasli was intended as a simple device; any backplane support is more secondary (if not feature creep).

**Comment by hartytp** [comment]
> @gkasprow: We decided to go for IDCs to enable out of rack operation of Kasli extensions.

Agreed, we definitely want to keep the IDCs. The questions was whether we have the backplane as well and, if so, what we are trying to achieve with it.

> @gkasprow: IDCs can be low or high performance. You can easily run 2.5Gbit (PCIe) data over such cable so there should be no issues with DDS modules.

Agreed, signal integrity for LVDS over ribbon cable should not be an issue. 

> If we switch to backplane connectivity we would have to develop dedicated custom backplane for every combination of extension modules. Keep in mind that these modules may have different size so there will always be wasted some slots on the backplane.

Agreed, if we don't constrain the module sizes better then the backplane topology gets to be a mess. This is something that would need to be thought about 

> @sbourdeauducq: Also Kasli was intended as a simple device; any backplane support is more secondary (if not feature creep).

Agreed, backplane support is secondary and definitely not required. But, as @dtcallcock points out, without it the systems can start to descend into a rats' nest of ribbon cable.

FWIW, I don't feel strongly about whether we have a backplane or not. I'm mainly just trying to clarify the various suggestions that have been put forward and ensure that people aren't talking cross purposes.

I would like to make sure that if we do have a BP, that it's properly thought through and has a well-documented motivation. If we're not going to do that then I don't see any point in putting the 96-pin DIN connector on the Kasli/VHDCI boards in the first place.

**Comment by hartytp** [comment]
> @jbqubit: Please update Kasli specification on wiki to reflect details fleshed out in this Issue.

Done. Although, let's not worry too much about the details of this (e.g. 3 v 4 SPF) until someone actually gets funding for it.

**Comment by jbqubit** [comment]
@hartytp Thanks for the cleanup of the wiki and update of Kasli specification. 

Comments on recent discussion:
- The EEM peripherals are about doing simple things simply. The IDC connector for EEM is simple and should be used for our initial round of 3U peripherals. The "rat's nest" is unlikely to be unruly by physicist standards since EEM consolidates IO, I2C, and power. 
- The Kasli Carrier and VHDCI Carrier should be able to interchangeably drive EEM peripherals. 
- Extend EEM specification to add an extra pin pair for differential clock.  2x15 -> 2x16  
- I don't see a strong use case for a backplane. I advocate that we drop the backplane from the specification entirely. Let future users define this. This also permits the use of a smaller FGPA -- see below. 

I've taken a quick look at the number of IO used by the current design. Looks like at least XC7A75T is required if we support 8 IDC and 4 backplane EEM. A XC7A75T may suffice if we forgo the back plane.  [Google doc](https://docs.google.com/spreadsheets/d/1iN-jyAVFPC8MUDF9MQv1YPmWOJgsYJULz3CoXzkRaJ4/edit?usp=sharing) is open to edit. 

<img width="432" alt="capture" src="https://cloud.githubusercontent.com/assets/20543958/23175417/525cece2-f82d-11e6-807e-7d18c9b9d830.PNG">

**Comment by jordens** [comment]
To repeat what @sbourdeauducq said: If we do this (we plan to do so), Kasli will be simple and cheap. Backplane support will be there but probably just secondary. The backplane might -- for example --- end up being just an alternative to using the first four IDC connectors. I.e. you can still only connect 8 extensions. But you can use the backplane to conveniently (mechanically and electrically) wire the first four extensions. We see no problem with VHDCI breakout compatibility but where adequate we may choose to give that a low priority. 
@jbqubit, please undo your sprinkling of "EEM" everywhere. Inventing new and wrong acronyms is not helpful. None of the extensions are Eurocards at this moment. And currently I see no reason for more than a 50T. Your pin count is off and not how we prioritize of even approach this. I don't understand what you are trying to achieve.
Instead it would great if you could contribute to publication quality datasheets for Sayma and Metlino (actual numbers, features, configuration options, high quality block diagrams) that detail the specification and convince other potential users of your design decisions and of the hardware quality.

**Comment by jordens** [comment]
Well. The EEM may have been @hartytp. Anyway, AFAICS most of them are not 100+n*60 mm long...

**Comment by gkasprow** [comment]
@filipswit started seriously working on Sayma and Metlino documentation. Filip just finished schemtic review of all boards and found several issues. He will use tex templates from AFCK.

**Comment by hartytp** [comment]
> @gkasprow: @filipswit started seriously working on Sayma and Metlino documentation. Filip just finished schemtic review of all boards and found several issues. He will use tex templates from AFCK.

What do people think about the idea of using the Wiki as the main source of documentation? This could have a few advantages:
* Easy for people to edit later on, for example to post measurement results or comment on hardware bugs we find
* Allows us to keep the entire Sinara ecosystem documented in a single place, rather than having a tex file for each component
* It's already a git repo, which makes version control easy

**Comment by jbqubit** [comment]
@hartytp @gkasprow Please move discussion on documentation to #133.

**Comment by hartytp** [comment]
@jordens The EEM terminology and pinout were mine. Apologies if I've made mistakes here, or if you feel this was unhelpful.

To be clear about my motivation for the updates I've made to the Wiki: we intend to fund the DAC and ADC boards imminently (hopefully in the next couple of weeks). The plan is to make something that we can run from the VHDCI carrier in the short term, but which will ultimately work with Kasli. From reading the issues, I get the impression that there are various conflicting ideas about what this hardware will do, and how it should fit together. I'd like to get a coherent, well documented plan before we start building hardware -- for example, if there's going to be a backplane then we will make our DACs compatible with it, but we can't do that until the specification is written down somewhere. If you have strong opinions, feel free to put them on the Wiki and we'll be happy to accommodate them.

Turning to your more specific points:

> None of the extensions are Eurocards at this moment.

You're correct, my language was imprecise. I mean that the boards should fit *within* the Eurocard form-factor, so that all boards are 100mm wide and no deeper than 160mm. AFAIK all currently planned boards do fit this. I'll amend the Wiki.

> please undo your sprinkling of "EEM" everywhere. Inventing new and wrong acronyms is not helpful. 

When I've had conversations with people about this, I found that I needed a convenient short-hand way of referring to this hardware (and to differentiate it from the uTCA sub-ecosystem). EEM emerged as an obvious convenient choice. I think it's important (and certainly not unhelpful) to have *some* name for them. If you have a better name then please introduce it.

> Your pin count is off and not how we prioritize of even approach this. I don't understand what you are trying to achieve.

Which pin count? Do you mean the 2x15 pin header pin count? That was taken from the current VHDCI/3U_BNC designs. What mistake have I made? My goal is to have a pinout that I can design our DAC/ADC boards around.

> Backplane support will be there but probably just secondary. The backplane might -- for example --- end up being just an alternative to using the first four IDC connectors. ... We see no problem with VHDCI breakout compatibility but where adequate we may choose to give that a low priority.

Can we at least try to agree on specifications (pinout etc) for the headers + backplane now? We plan to start building hardware based on the current VHDCI carrier design soon. It seems perverse to start building a Kasli carrier in a few months that won't work with DAC/ADC extensions we start building now.

> I.e. you can still only connect 8 extensions. But you can use the backplane to conveniently (mechanically and electrically) wire the first four extensions. 

> The backplane might -- for example --- end up being just an alternative to using the first four IDC connectors. I.e. you can still only connect 8 extensions. But you can use the backplane to conveniently (mechanically and electrically) wire the first four extensions. 

If you have strong opinions about this, please add them to the Wiki, and we'll design our hardware around them.

**Comment by hartytp** [comment]
Also, a small bugbear: in a few places, people are calling the 100mil male pin header on these boards "IDC". IDC (insulation-displacement contact) is a method of crimping ribbon cable onto housings. These housings mate with the pin headers on the board, but are not the connectors used on the boards themselves. Can we stick to "pin header" or similar please to avoid confusing things?

**Comment by gkasprow** [comment]
The 3U boards have such pin headers surrounded by walls which are mating parts for  crimped ribbon cable IDC connectors.
So we pin headers are like this:
![obraz](https://cloud.githubusercontent.com/assets/4325054/23190969/762db25a-f89b-11e6-9e53-1b5a3e00e808.png)
while IDC headers are like that:
![obraz](https://cloud.githubusercontent.com/assets/4325054/23190997/98aadf42-f89b-11e6-8d81-a977247dc3c1.png)

**Comment by hartytp** [comment]
@gkasprow Okay, I'm used to seeing that referred to a "shrouded" or "boxed" pin header, rather than an "IDC" header. But, if this is standard terminology then I'm wrong and I rescind my objections. Live and learn ;)

**Comment by gkasprow** [comment]
Whatever you google it gives the same pics :)
But indeed, shrouded is official name in many producers catalogues.
But since IDC header means that you plug IDC jumper we can leave this for
clarity.

**Comment by hartytp** [comment]
ack

**Comment by gkasprow** [comment]
To define backplane structure we first have to think of communication channel for backplane enabled devices. IMHO we should reserve backplane option for special use cases. Individual IO channels like SMA, BNC, RJ45 need dedicated LVDS links and IDCs do the job nicely. Backplane enabled modules will benefit from clock distribution so we can install there DDS, RF generators, DACs, ADCs that need external synchronization.
For backplane devices we can make dedicated VHDCI adapter that supports high speed MLVDS.
So we can define backplane as MLVDS bus with limited number of supported protocols.
In this way all modules may have identical pinout and routed i.e. 16 LVDS lines + one dedicated selection line.
The interface that fits best both simple and complex devices is SPI over LVDS with multiple incarnations like QSPI, double QSPI. It resembles MMC interface where clock, chip select and 8 data lines are used.
With such interface running across the backplane, we can insert as many devices as the rack fits and obtain scalable transfer rate. It has one more benefit - we can mix 4HP and 8HP modules without loosing too many Kasli IO lines - only single chip select is not used.
 For simple DACs we simply use standard SPI and ignore remaining data lines.
For applications that require hundreds of MB/s transfer rate (high speed ADC, DAC modules) we simply use 8 or 16 bit interface. To make all deterministic we can define time slots for every device.
SPI clock speed can also vary depending on currently selected module. MLVDS supports up to 100MHz.

**Comment by jordens** [comment]
Rough Kasli design looks settled. Backplane discussion moved to #168.
